File: ejemplos/compo1.clj
(ns ejemplos.compo1)

(defn sumar [x y] (+ x y))
(defn duplicar [x] (* x 2))

;; En clojure hay tres maneras de componer estas funciones
(defn sumar-y-duplicar [x y] (duplicar (sumar x y)))
(sumar-y-duplicar 1 2)

(def sumar-y-duplicar2 (comp duplicar sumar))
(sumar-y-duplicar2 1 2)

(defn sumar-y-duplicar3 [x y] (-> (sumar x y) duplicar ))
(sumar-y-duplicar3 1 2)

File: ejemplos/compo1.hs
sumar:: (Int -> Int) -> Int
sumar x y = x + y

duplicar:: (Int) -> Int
duplicar x = x * 2

-- En Haskell hay dosmaneras de componer estas funciones
sumarYDuplicar :: Int -> Int -> Int
sumarYDuplicar x y = duplicar (sumar x y)

sumarYDuplicar :: Int -> Int -> Int
sumarYDuplicar2 = duplicar . sumar
